The dataset contains an ID column, 25 features and one target label column. The description of the features is given below:

    url : URL of the webpage to be classified
    webpageDescription : One line description of the webpage
    alchemy_category : Alchemy category (per the publicly available Alchemy API)
    alchemy_category_score : Alchemy category score (per the publicly available Alchemy API)
    avgLinkWordLength : Average number of words in a webpage
    AvglinkWithOneCommonWord : Average number of web pages sharing at least one word with one other web page
    AvglinkWithTwoCommonWord : Average number of web pages sharing at least one word with two other web pages
    AvglinkWithThreeCommonWord : Average number of web pages sharing at least one word with three other web pages
    AvglinkWithFourCommonWord : Average number of web pages sharing at least one word with four other web pages
    redundancyMeasure : Measure of redundancy computed by finding the compression achieved on this web page via gzip
    embedRatio : Count of tags
    frameBased : Binary indication of whether a webpage has frameset markup
    frameTagRatio : Ratio of frameset markups over total markups
    domainLink: Binary indication of whether the webpage contains in URL with domain
    tagRatio : Ratio of tags over text in the webpage
    imageTagRatio : Ratio of tags over text in the webpage
    isNews : Binary indication of whether a webpage is a news article
    lengthyDomain : Binary indication of whether webpage's text contains more than 30 alpha-numeric characters
    hyperlinkToAllWordsRatio : Percentage of words on the webpage that are also in the hyperlink text
    isFrontPageNews : Binary indication of whether webpage is front-page news
    alphanumCharCount : Number of alpha-numeric characters in webpage's text
    linksCount : Number of markups
    wordCount : Number of words in URL
    parametrizedLinkRatio :
    spellingErrorsRatio : Ration of words that contain spelling errors

The label value of 0 represents that the webpage is not "ad-worthy", and a label value of 1 represents that the webpage is "ad-worthy". Your task is to predict the label value based on the features described above.

In [16]:
import numpy as np
import pandas as pd
import json

In [17]:
train_data = pd.read_csv('dataset/train_data.csv')
test_data = pd.read_csv('dataset/test_data.csv')

### Duplicate rows

In [18]:
print(train_data['url'].duplicated().sum())
print(test_data.duplicated().sum())

0
0


No duplicate rows

### Merging train and test data

### Converting webpageDescription from string to JSON data

In [19]:
merged_data = pd.concat([train_data, test_data])

merged_data['webpageDescription'] = merged_data['webpageDescription'].apply(lambda x: json.loads(x))

### Checking missing entries in each column

In [20]:
missing_entries = []

for feature in merged_data.columns:
    record = {}
    record['feature'] = feature
    record['NA values'] = merged_data[feature].isna().sum()
    record['? values'] = len(merged_data[merged_data[feature] == '?'][feature])
    missing_entries.append(record)

pd.DataFrame(missing_entries)

feature  NA values  ? values
0                          url          0         0
1           webpageDescription          0         0
2             alchemy_category          0      2342
3       alchemy_category_score          0      2342
4            avgLinkWordLength          0         0
5     AvglinkWithOneCommonWord          0         0
6     AvglinkWithTwoCommonWord          0         0
7   AvglinkWithThreeCommonWord          0         0
8    AvglinkWithFourCommonWord          0         0
9            redundancyMeasure          0         0
10                  embedRatio          0         0
11                  framebased          0         0
12               frameTagRatio          0         0
13                  domainLink          0         0
14                    tagRatio          0         0
15               imageTagRatio          0         0
16                      isNews          0      2843
17               lengthyDomain          0         0
18    hyperlinkToAllWordsRatio          0         0
19             isFrontPageNews          0      1248
20           alphanumCharCount          0         0
21                  linksCount          0         0
22                   wordCount          0         0
23       parametrizedLinkRatio          0         0
24         spellingErrorsRatio          0         0
25                       label       1479         0
26                          id          0         0

### AlchemyCategory & AlchemyCategoryScore has 2342 missing entries (? values)

### isNews 2843 entries (? values)

### isFrontpageNews 1248 entries (? values)



### Analyzing the keys in JSON value of webpageDescription

It seems that "body" key is the one-line description of the page as described in the dataset description

In [21]:
merged_data['webpageDescription'].apply(lambda x: x.keys())

0       (title, body, url)
1            (body, title)
2       (title, body, url)
3       (title, body, url)
4       (url, title, body)
               ...        
1474    (title, body, url)
1475    (title, body, url)
1476    (url, title, body)
1477    (title, body, url)
1478    (title, body, url)
Name: webpageDescription, Length: 7395, dtype: object

In [22]:
len(merged_data[merged_data['webpageDescription'].apply(lambda x: x['body'] == None)])

57

### There are 57 entries where the webpageDescription has no "body" entry

In [23]:
merged_data[merged_data['webpageDescription'].apply(lambda x: x['title'] if x['body'] == None else x['body']).isna()]

url  \
2994  http://icanhascheezburger.com/2009/05/07/funny...   

                                     webpageDescription alchemy_category  \
2994  {'title': None, 'body': None, 'url': 'icanhasc...                ?   

     alchemy_category_score  avgLinkWordLength  AvglinkWithOneCommonWord  \
2994                      ?           1.982178                  0.756809   

      AvglinkWithTwoCommonWord  AvglinkWithThreeCommonWord  \
2994                  0.285992                    0.235409   

      AvglinkWithFourCommonWord  redundancyMeasure  ...  lengthyDomain  \
2994                   0.210117                0.0  ...              1   

      hyperlinkToAllWordsRatio  isFrontPageNews  alphanumCharCount  \
2994                        35                ?               8682   

      linksCount  wordCount parametrizedLinkRatio  spellingErrorsRatio  label  \
2994         514          6              0.206226                0.125    0.0   

        id  
2994  5330  

[1 rows x 27 columns]

### There is 1 entry where the webpageDescription has no "body" entry and no "title" entry too

This entry belongs to train set, so we can just delete it

### Replacing webpageDescription using the following logic

    All entries have "body" key in webpageDescription column but some have these values as None
    So we fill in those entries with "title" key value
    
    if webpageDescription["body"] != None
        webpageDescription["body"]
    else webpageDescription["title"]

In [24]:
merged_data[merged_data['webpageDescription'].apply(lambda x: x['title'] if x['body'] == None else x['body']).isna()]

cleaned_data = merged_data.copy(deep=True)
cleaned_data.drop(index=2994, inplace=True)

In [25]:
cleaned_data['webpageDescription'] = cleaned_data['webpageDescription'].apply(lambda x: x['title'] if x['body'] == None else x['body'])

### Analyzing value_counts() of the two news columns

In [26]:
cleaned_data['isNews'].value_counts()

1    4552
?    2842
Name: isNews, dtype: int64

It seems pretty reasonable to assume that ? refers to isNews = 0 but more analysis needs to be done

In [27]:
cleaned_data['isFrontPageNews'].value_counts()

0    5853
?    1247
1     294
Name: isFrontPageNews, dtype: int64

It also seems reasonable to assume that ? refers to isFrontpageNews = 1 but more analysis needs to be done

In [28]:
cleaned_data[cleaned_data['isNews'] == '?'].head(10)

url  \
0         http://www.polyvore.com/cgi/home?id=1389651   
1          http://www.youtube.com/watch?v=ippMPPu6gh4   
4              http://recipes.wuzzle.org/index.php/72   
6   http://www.insidershealth.com/natural_cure/poi...   
7                         http://yawoot.com/post/2122   
16  http://www.nationalpost.com/news/story.html?id...   
17          http://thecompletecookbook.wordpress.com/   
21              http://www.pimpthatsnack.com/full.php   
22  http://www.etsy.com/treasury/MTA0NjAxODZ8NzkwM...   
24            http://warehouse.carlh.com/article_157/   

                                   webpageDescription    alchemy_category  \
0    polyvore is the best place to discover or sta...                   ?   
1   Speed Air Man--David Belle david belle speed a...                   ?   
4   Barbecued Chicken Chow Siew from The Exotic Ki...   computer_internet   
6   Copyright 2011 InsidersHealth com All Rights R...              health   
7   Well the world is made of peanut butter now po...          recreation   
16  In a discovery that has stunned even those beh...              health   
17  Before we get to today s post please understan...  science_technology   
21  Pimp That Snack Snack Search Join the mailing ...    culture_politics   
22                                                       culture_politics   
24  February 4th 2008 Copy the line below to link ...  arts_entertainment   

   alchemy_category_score  avgLinkWordLength  AvglinkWithOneCommonWord  \
0                       ?           1.916667                  0.047619   
1                       ?           1.257576                  0.141026   
4                0.535009           0.181818                  0.036364   
6                0.893828           1.691781                  0.469799   
7                0.550131           1.170213                  0.634921   
16               0.603181           3.261649                  0.653595   
17               0.564931           1.371717                  0.475904   
21               0.208574           2.388489                  0.733813   
22               0.165873           1.488235                  0.408451   
24               0.361629           2.000000                  0.086957   

    AvglinkWithTwoCommonWord  AvglinkWithThreeCommonWord  \
0                   0.007937                    0.000000   
1                   0.000000                    0.000000   
4                   0.000000                    0.000000   
6                   0.147651                    0.026846   
7                   0.000000                    0.000000   
16                  0.369281                    0.196078   
17                  0.269076                    0.128514   
21                  0.496403                    0.302158   
22                  0.140845                    0.014085   
24                  0.000000                    0.000000   

    AvglinkWithFourCommonWord  redundancyMeasure  ...  lengthyDomain  \
0                    0.000000           0.803797  ...              0   
1                    0.000000           1.142857  ...              0   
4                    0.000000           0.292614  ...              0   
6                    0.006711           0.522134  ...              0   
7                    0.000000           0.224906  ...              0   
16                   0.140523           0.442116  ...              1   
17                   0.054217           0.511271  ...              1   
21                   0.086331           0.723183  ...              0   
22                   0.000000          21.000000  ...              0   
24                   0.000000           0.485349  ...              0   

    hyperlinkToAllWordsRatio  isFrontPageNews  alphanumCharCount  linksCount  \
0                         34                0                682         126   
1                         12                0               3008          78   
4                          3                0               1745 

In [29]:
cleaned_data.iloc[16, 1]

'In a discovery that has stunned even those behind it scientists at a Toronto hospital say they have proof the body s nervous system helps trigger diabetes opening the door to a potential near cure of the disease that affects millions of Canadians Diabetic mice became healthy virtually overnight after researchers injected a substance to counteract the effect of malfunctioning pain neurons in the pancreas I couldn t believe it said Dr Michael Salter a pain expert at the Hospital for Sick Children and one of the scientists Mice with diabetes suddenly didn t have diabetes any more The researchers caution they have yet to confirm their findings in people but say they expect results from human studies within a year or so Any treatment that may emerge to help at least some patients would likely be years away from hitting the market But the excitement of the team from Sick Kids whose work is being published today in the journal Cell is almost palpable I ve never seen anything like it said Dr 

It is clear that this is a news article and its 'isNews' value was set to '?'

This just tells us that we'll have to closely analyze each entry with '?' for isNews to determine whether they're a news page or not

Or we can consider this to be an outlier and assume that its mostly non-news page whenever isNews = '?'

A separate NLP model can be trained to predict whether that page isNews or not given the description of the page. But this can be very unreliable plus we don't have class label 0 entries for isNews in the given dataset. So it makes more sense to straight manually label each entry for whether it is news or not.

Similar manual labelling can be done for isFrontPageNews as well


These same trends were observed on the test data separately as well

### Analyzing isFrontPageNews column for ? values

In [30]:
cleaned_data[cleaned_data['isFrontPageNews'] == '?']['isNews'].value_counts()

?    1247
Name: isNews, dtype: int64

The entries which have ? values in isFrontPageNews also have ? values in isNews

### Analyzing isNews column for ? values

In [31]:
cleaned_data[cleaned_data['isNews'] == '?']['webpageDescription']

0        polyvore is the best place to discover or sta...
1       Speed Air Man--David Belle david belle speed a...
4       Barbecued Chicken Chow Siew from The Exotic Ki...
6       Copyright 2011 InsidersHealth com All Rights R...
7       Well the world is made of peanut butter now po...
                              ...                        
1471                                                     
1474                                                     
1475    Save to your Collections Sorry for the inconve...
1477    outerbanxchic posted 10 14 2011 Delicious and ...
1478    Summer always signifies the beginning of steak...
Name: webpageDescription, Length: 2842, dtype: object

### Analyzing the alchemy_category column

In [32]:
cleaned_data['alchemy_category'].value_counts()

?                     2341
recreation            1229
arts_entertainment     941
business               880
health                 506
sports                 380
culture_politics       343
computer_internet      296
science_technology     289
gaming                  76
religion                72
law_crime               31
unknown                  6
weather                  4
Name: alchemy_category, dtype: int64

In [33]:
cleaned_data[cleaned_data['alchemy_category'] == 'law_crime'].head()

url  \
66    http://vii2012.com/2012/06/accidental-90s-nick...   
377   http://www.independent.co.uk/news/world/americ...   
639   http://newsfeed.time.com/2012/01/09/reading-wh...   
847   http://www.insidethekaganoffkitchen.com/2010/0...   
1334  http://www.magnoliarouge.com/2012/08/outfit-of...   

                                     webpageDescription alchemy_category  \
66    Your smiling at me is my daily dose of magic S...        law_crime   
377   google ad client ca pub 5964551156905038 if re...        law_crime   
639   Monday s links talk surprising birthdays body ...        law_crime   
847   by Rachel on April 11 2010 I just love a good ...        law_crime   
1334  AboutWelcome to Magnolia Rouge a collection of...        law_crime   

     alchemy_category_score  avgLinkWordLength  AvglinkWithOneCommonWord  \
66                 0.505332           3.666667                  0.545455   
377                0.818603           2.068460                  0.579572   
639                0.752151           2.972222                  0.553333   
847                0.246582           1.904255                  0.392523   
1334               0.313761           1.300971                  0.258929   

      AvglinkWithTwoCommonWord  AvglinkWithThreeCommonWord  \
66                    0.454545                    0.318182   
377                   0.187648                    0.054632   
639                   0.193333                    0.100000   
847                   0.242991                    0.121495   
1334                  0.071429                    0.035714   

      AvglinkWithFourCommonWord  redundancyMeasure  ...  lengthyDomain  \
66                     0.227273           0.621094  ...              0   
377                    0.033254           0.432215  ...              1   
639                    0.073333           0.504981  ...              1   
847                    0.112150           0.522953  ...              1   
1334                   0.017857           0.675141  ...              0   

      hyperlinkToAllWordsRatio  isFrontPageNews  alphanumCharCount  \
66                          51                0                306   
377                         53                0               3828   
639                         57                0               1595   
847                         19                0               3475   
1334                        45                0               1034   

      linksCount  wordCount parametrizedLinkRatio  spellingErrorsRatio  label  \
66            22          4              0.000000             0.161290    0.0   
377          421         14              0.076010             0.093960    1.0   
639          150          8              0.126667             0.087108    0.0   
847          107          2              0.009346             0.089928    1.0   
1334         112          4              0.241071             0.136986    0.0   

        id  
66    6771  
377   3319  
639    452  
847   3203  
1334  1628  

[5 rows x 27 columns]

### This cell actually shows you how the alchemy_category can be misclassified

Because some of these pages are clearly not related to law_crime category

This tells us that alchemy_category and the corresponding score cannot be relied on too much and perhaps we can try thresholding off the basis of alchemy_category_score, where we only retain those categories where the score is high, like 60%



In [34]:
cleaned_data[cleaned_data['alchemy_category_score'].apply(lambda x: 0 if x == '?' else float(x)) > 0.6]

url  \
5     http://www.sleepdisordersguide.com/blog/good-n...   
6     http://www.insidershealth.com/natural_cure/poi...   
12    http://sugarcrafter.net/2009/08/20/sesame-chic...   
16    http://www.nationalpost.com/news/story.html?id...   
19    http://bleacherreport.com/articles/1319130-100...   
...                                                 ...   
1472  http://www.deepsouthdish.com/2010/10/macaroni-...   
1473  http://www.nanny.net/blog/18-blogs-featuring-t...   
1474  http://www.lanascooking.com/2011/07/08/roasted...   
1476  http://www.laweekly.com/bestof/2011/award/best...   
1478                        http://canarygirl.com/?p=80   

                                     webpageDescription    alchemy_category  \
5     You have to pay high price if you are not gett...              health   
6     Copyright 2011 InsidersHealth com All Rights R...              health   
12    August 20 2009 Print E mail Filed under Asian ...  arts_entertainment   
16    In a discovery that has stunned even those beh...              health   
19    100 Ashley Judd Dario Franchitti Superfan Poli...  arts_entertainment   
...                                                 ...                 ...   
1472  As soon as I saw this Mac n Cheese Soup recipe...          recreation   
1473  Posted on June 5 2013 by admin in NannyMillion...              health   
1474                                                               business   
1476  Best Belly Dance Workout 2011 Swerve Studio s ...  arts_entertainment   
1478  Summer always signifies the beginning of steak...  arts_entertainment   

     alchemy_category_score  avgLinkWordLength  AvglinkWithOneCommonWord  \
5                  0.953935           3.697917                  0.848485   
6                  0.893828           1.691781                  0.469799   
12                 0.722537           2.573529                  0.451220   
16                 0.603181           3.261649                  0.653595   
19                     0.85           2.649573                  0.496503   
...                     ...                ...                       ...   
1472                0.65858           1.125320                  0.295122   
1473               0.625404           4.294737                  0.845361   
1474               0.681418           1.640000                  0.367521   
1476               0.943693           1.829787                  0.512690   
1478               0.715035           2.020243                  0.491935   

      AvglinkWithTwoCommonWord  AvglinkWithThreeCommonWord  \
5                     0.636364                    0.191919   
6                     0.147651                    0.026846   
12                    0.195122                    0.146341   
16                    0.369281                    0.196078   
19                    0.237762                    0.118881   
...                        ...                         ...   
1472                  0.151220                    0.078049   
1473                  0.701031                    0.360825   
1474                  0.102564                    0.034188   
1476                  0.279188                    0.055838   
1478                  0.217742                    0.032258   

      AvglinkWithFourCommonWord  redundancyMeasure  ...  lengthyDomain  \
5                      0.101010           0.382585  ...              1   
6                      0.006711           0.522134  ...              0   
12                     0.097561           0.487435  ...              1   
16                     0.140523           0.442116  ...              1   
19                     0.055944           0.463007  ...              1   
...                         ...                ...  ...            ...   
1472                   0.043902           0.482424  ...              1   
1473                   0.144330           0.466958  ...              1   
1474                   0.017094          21.000000  ...              1

There are in total 2733 rows where we have alchemy_category_score > 0.6, which means a 60% confidence in the fact that the alchemy_category is correct. 

In total we have 7394 entries out of which 2341 are ? entries and 2733 entries with high confidence, so we have 7394 - 2341 - 2733 = 2320 entries with tagged alchemy_category but low confidence

In [35]:
cleaned_data[cleaned_data['alchemy_category'] == '?']

url  \
0           http://www.polyvore.com/cgi/home?id=1389651   
1            http://www.youtube.com/watch?v=ippMPPu6gh4   
11    http://www.chow.com/recipes/13499-creamy-carro...   
15    http://www.whattoexpect.com/preconception/heal...   
23    http://www.washingtonpost.com/lifestyle/wellne...   
...                                                 ...   
1467  http://www.healthguru.com/applications/quiz/qz...   
1469  http://news.menshealth.com/touch-at-your-own-p...   
1471                          http://dentalcalgary.org/   
1475  http://www.marthastewart.com/255456/billys-cho...   
1477  http://www.kraftrecipes.com/recipes/buffalo-ch...   

                                     webpageDescription alchemy_category  \
0      polyvore is the best place to discover or sta...                ?   
1     Speed Air Man--David Belle david belle speed a...                ?   
11    Difficulty Easy Total Time 55 mins Active Time...                ?   
15    Next time you re in the bathroom crack open yo...                ?   
23    The oil is the healthiest part of a nut contai...                ?   
...                                                 ...              ...   
1467  Weird Facts on Drugs Quiz HealthGuru Question ...                ?   
1469  It s a wonder your hand doesn t melt off every...                ?   
1471                                                                   ?   
1475  Save to your Collections Sorry for the inconve...                ?   
1477  outerbanxchic posted 10 14 2011 Delicious and ...                ?   

     alchemy_category_score  avgLinkWordLength  AvglinkWithOneCommonWord  \
0                         ?           1.916667                  0.047619   
1                         ?           1.257576                  0.141026   
11                        ?           1.804000                  0.436090   
15                        ?           2.476562                  0.558824   
23                        ?           2.919403                  0.632312   
...                     ...                ...                       ...   
1467                      ?           2.243094                  0.622951   
1469                      ?           0.000000                  0.000000   
1471                      ?           1.363636                  0.230769   
1475                      ?           2.236559                  0.676768   
1477                      ?           1.743316                  0.559091   

      AvglinkWithTwoCommonWord  AvglinkWithThreeCommonWord  \
0                     0.007937                    0.000000   
1                     0.000000                    0.000000   
11                    0.086466                    0.007519   
15                    0.250000                    0.095588   
23                    0.281337                    0.130919   
...                        ...                         ...   
1467                  0.327869                    0.120219   
1469                  0.000000                    0.000000   
1471                  0.153846                    0.000000   
1475                  0.232323                    0.090909   
1477                  0.227273                    0.081818   

      AvglinkWithFourCommonWord  redundancyMeasure  ...  lengthyDomain  \
0                      0.000000           0.803797  ...              0   
1                      0.000000           1.142857  ...              0   
11                     0.003759           0.476190  ...              1   
15                     0.014706           0.450391  ...              1   
23                     0.077994           0.440646  ...              1   
...                         ...                ...  ...            ...   
1467                   0.027322           0.858025  ...              1   
1469                   0.000000           0.512821  ...              0   
1471                   0.000000          21.000000  ...              0   
1475                   0.020202 

Similar to the isNews labelling, the ? values can fall into any one of the given categories which can be inferred pretty quickly given the context. This can either be manually filled in or a separate NLP classifier can be created.

It also makes sense that once we have finished labelling, the alchemy_category_score column doesn't make sense as by that point we'll only leave those entries in alchemy_category for which we have very high confidence

### Analyzing the numerical columns in the dataset

Basically analyzing the output of the describe() call on the dataset

In [36]:
merged_data.describe()

avgLinkWordLength  AvglinkWithOneCommonWord  AvglinkWithTwoCommonWord  \
count        7395.000000               7395.000000               7395.000000   
mean            2.761823                  0.468230                  0.214080   
std             8.619793                  0.203133                  0.146743   
min             0.000000                  0.000000                  0.000000   
25%             1.602062                  0.340370                  0.105263   
50%             2.088235                  0.481481                  0.202454   
75%             2.627451                  0.616604                  0.300000   
max           363.000000                  1.000000                  1.000000   

       AvglinkWithThreeCommonWord  AvglinkWithFourCommonWord  \
count                 7395.000000                7395.000000   
mean                     0.092062                   0.049262   
std                      0.095978                   0.072629   
min                      0.000000                   0.000000   
25%                      0.022222                   0.000000   
50%                      0.068627                   0.022222   
75%                      0.133333                   0.065065   
max                      0.980392                   0.980392   

       redundancyMeasure   embedRatio  framebased  frameTagRatio   domainLink  \
count        7395.000000  7395.000000      7395.0    7395.000000  7395.000000   
mean            2.255103    -0.103750         0.0       0.056423     0.021231   
std             5.704313     0.306545         0.0       0.041446     0.144162   
min             0.000000    -1.000000         0.0       0.000000     0.000000   
25%             0.442616     0.000000         0.0       0.028502     0.000000   
50%             0.483680     0.000000         0.0       0.045775     0.000000   
75%             0.578227     0.000000         0.0       0.073459     0.000000   
max            21.000000     0.250000         0.0       0.444444     1.000000   

       ...  imageTagRatio  lengthyDomain  hyperlinkToAllWordsRatio  \
count  ...    7395.000000    7395.000000               7395.000000   
mean   ...       0.275709       0.660311                 30.077079   
std    ...       1.919320       0.473636                 20.393101   
min    ...      -1.000000       0.000000                  0.000000   
25%    ...       0.025900       0.000000                 14.000000   
50%    ...       0.083051       1.000000                 25.000000   
75%    ...       0.236700       1.000000                 43.000000   
max    ...     113.333333       1.000000                100.000000   

       alphanumCharCount   linksCount    wordCount  parametrizedLinkRatio  \
count        7395.000000  7395.000000  7395.000000            7395.000000   
mean         5716.598242   178.754564     4.960649               0.172864   
std          8875.432430   179.466198     3.233111               0.183286   
min             0.000000     1.000000     0.000000               0.000000   
25%          1579.000000    82.000000     3.000000               0.040984   
50%          3500.000000   139.000000     5.000000               0.113402   
75%          6377.000000   222.000000     7.000000               0.241299   
max        207952.000000  4997.000000    22.000000               1.000000   

       spellingErrorsRatio        label           id  
count          7395.000000  5916.000000  7395.000000  
mean              0.101221     0.517241  3697.000000  
std               0.079231     0.499745  2134.896953  
min               0.000000     0.000000     0.000000  
25%               0.068739     0.000000  1848.500000  
50%               0.089312     1.000000  3697.000000  
75%               0.112376     1.000000  5545.500000  
max               1.000000     1.000000  7394.000000  

[8 rows x 21 columns]

avgLinkWordLength has a max value of 363 which is a bit outlier-ish considering both the median and mean are around the 2-3 mark.

However avgLinkWordLength measures the average number of words in a webpage and its sensible that a particular page can have 363 words in it. So you cannot consider it as an invalid value or such.

Simple outlier fixing like setting the outlier values to the Q3 quartile is also valid.

Regardless, it will be beneficial if feature standardization is done on all these numeric columns.

embedRatio column is described to contain the "count of tags" but the column has negative values so it becomes difficult to interpret what "count of tags" refers to. It is important to note that the max value in the column is 0.25 which does seem to indicate that this is a low value column.

framebased seems to have just 0 values, so that can be dropped

domainLink (binary column) seems to be skewed toward the 0 class.

hyperlinkToAllWordsRatio (percentage column) has values in range 0 to 100, so erroneous or outlier values there.

alphanumCharCount understandly has very high values. However min value is 0 but this is also possible because some webpages in the dataset have only images or videos hosted on them.

linksCount, wordCount, parameterizedLinkRatio and spellingErrorsRatio also seem to have appropriate values


In [37]:
 merged_data['framebased'].value_counts()

0    7395
Name: framebased, dtype: int64

The framebased column is completely 0. So we can just drop it

In [38]:
merged_data.drop('framebased', axis=1, inplace=True)

### Analyzing tagRatio column

In [41]:
merged_data['tagRatio'].describe()

count    7395.000000
mean        0.233778
std         0.052487
min         0.045564
25%         0.201061
50%         0.230564
75%         0.260770
max         0.716883
Name: tagRatio, dtype: float64

Everything looks alright here

### Analyzing domainLink column

In [39]:
merged_data['domainLink'].value_counts()

0    7238
1     157
Name: domainLink, dtype: int64

In [40]:
test_data['domainLink'].value_counts()

0    1436
1      43
Name: domainLink, dtype: int64

Huge class imbalance in both train and test data, but oversampling doesn't make much sense. At this point it is more sensible to straight up drop the column entirely but this can be tested later. Train model using this column and without using this column.

### Analyzing lengthyDomain column

In [42]:
merged_data['lengthyDomain'].value_counts()

1    4883
0    2512
Name: lengthyDomain, dtype: int64

Everything looks alright here